### PR TITLE
added undo to "Change Node Type"

### LINKED ISSUE #15936: "Change Type" no Changed * on Scene Tab
Godot Master

peek 2018-01-21 12-27

Scene Tab is not showing the * sign.
Undo "Change Type" is also not working.

https://github.com/godotengine/godot/blob/35700921f82d72525555947866ff283f833fee7d/editor/scene_tree_dock.cpp#L1699

### PR DESCRIPTION
**Pull request by AlexHolly**
closes https://github.com/godotengine/godot/issues/15936

peek 2018-08-24 13-44

<details>
  <summary> EDIT: Can't reproduce anymore</summary>
Need to double check something

```
ERROR: get_node: Node not found: /root/EditorNode/@@5/@@6/@@14/@@16/@@20/@@24/@@25/@@26/@@42/@@43/@@52/@@53/@@5414/@@5281/@@5282/@@5283/@@5284/@@5285/Control/Control3
   At: scene/main/node.cpp:1304.
ERROR: get_meta: Condition ' !metadata.has(p_name) ' is true. returned: Variant()
   At: core/object.cpp:1084.
test: Null
```
Restarting the Editor lead to a missing Node Icon.

</details>

--
Added Live debug and noticed that "Make Scene Root" and "Rename Node" have no live_debug

### CODE REVIEW

**Comment by Zylann** [comment]
How is the old node being freed?

**Comment by AlexHolly** [comment]
Yeah I asked my self the same thing. It is the same as `_delete_confirm` Hopefully when Godot closes.
https://github.com/godotengine/godot/blob/6d9cc032e71ccc768064bf908623b0954d01181c/editor/scene_tree_dock.cpp#L1520
https://github.com/godotengine/godot/blob/master/scene/main/node.cpp#L1169

File: editor/scene_tree_dock.cpp [hunk 1]
```diff
@@ -1684,20 +1684,39 @@ void SceneTreeDock::_create() {
...
 
-			replace_node(n, newnode);
+			String path_to_n = edited_scene->get_path_to(n);
+			int n_pos_in_parent = n->get_position_in_parent();
+			String n_name = n->get_name();
+
+			editor_data->get_undo_redo().add_do_method(this, "replace_node", n, new_node);
+			editor_data->get_undo_redo().add_do_reference(new_node);
+			editor_data->get_undo_redo().add_do_method(sed, "live_debug_replace_node", path_to_n, new_node->get_class(), n_name);
+			editor_data->get_undo_redo().add_do_method(this, "memdelete", n);
```
Comment (Xrayez) [inline editor/scene_tree_dock.cpp]: Not sure but `UndoRedo` should release the old node through reference passed by `add_do_reference/add_undo_reference` automatically if the undo history is full, I don't see using `do/redo` method for explicitly freeing a node anywhere throughout the engine, so perhaps they are redundant, and I'm not sure how do you keep/restore the node's state once you delete it.

**Comment by akien-mga** [comment]
Moving to 3.2 milestone as we're about to enter the beta stage and release freeze for Godot 3.1. Only bug fixes against issues of the 3.1 milestone (or enhancements to core features of the 3.1 roadmap) will be considered until 3.1-stable is released.

Note: If it's merged after 3.1-stable is released, this change might be considered for cherry-picking into a later 3.1.x release.

**Comment by aaronfranke** [comment]
What's the status of this PR? Why does the OP say can't reproduce?

**Comment by Jummit** [comment]
> What's the status of this PR? Why does the OP say can't reproduce?


He says that this isn't reproducable anymore:


> Need to double check something
ERROR: get_node: Node not found: /root/EditorNode/@@5/@@6/@@14/@@16/@@20/@@24/@@25/@@26/@@42/@@43/@@52/@@53/@@5414/@@5281/@@5282/@@5283/@@5284/@@5285/Control/Control3
   At: scene/main/node.cpp:1304.
ERROR: get_meta: Condition ' !metadata.has(p_name) ' is true. returned: Variant()
   At: core/object.cpp:1084.
test: Null

**Comment by akien-mga** [comment]
Seems like the original issue was fixed in 3.1 already, so this PR is no longer necessary.

**Pull request closed without merging**